### PR TITLE
Fix state machine memory leak

### PR DESCRIPTION
**Pull request by SeoFernando25**
- Removed unused platform specific code from stateMachineLib
- Created `pch.h` file ( [cmake](https://github.com/uorocketry/rocket-code-2020/pull/176/files#diff-1e7de1ae2d059d21e1dd75d5812d5a34b0222cef273b7c3a2af62eb747f9d20aR78), [pch.h](https://github.com/uorocketry/rocket-code-2020/pull/176/files#diff-74c786f33cde8f4f13642d5b25b2c7f4c9afda820e66daf55dfa924bfe3b845fR1) )
- Removed legacy type definitions in favor of `struct_t` types
- State machine's EventData is now explicitly passed by reference, therefore, it will never be null
    - The performance of it never being null is negligible compared to using raw pointers since the NoEventData is just an empty class
- No longer has to pass `nullptr` argument to `END_TRANSITION_MAP` macro

[Main sauce changes](https://github.com/uorocketry/rocket-code-2020/pull/176/files#diff-8e43c202ea7404f5e1470a227902a5d19f3f1a89ed6ebc16b95d1964a590026cL1)

Fixes: #72

### CODE REVIEW

File: src/common/pch.h [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+#include <memory>
+#include <mutex>
+#include <queue>
+#include <string>
+#include <thread>
+#include <unordered_map>
+#include <vector>
+
+// Logging
+#define SPDLOG_ACTIVE_LEVEL SPDLOG_LEVEL_TRACE
```
Comment (jonapap) [inline src/common/pch.h:19]: Might be able to remove this line. I think I added it a while ago when I couldn't get debug working, but from what I'm reading it shouldn't be needed. I'll have to check and test when I get back home.

Comment (jonapap) [reply src/common/pch.h:19]: So actually, moving this define to a global header fixes the log behaviour, which I haven't realized was even broken! Thanks! However, now the logs are maybe a bit too verbose :grin:. I'll push a commit to fix that.

File: CMakeLists.txt [hunk 2]
```diff
@@ -75,6 +75,8 @@ add_library(MainLoopLib ${SRC_MAIN} ${MAIN_CONFIG_SRC})
 
 target_include_directories(MainLoopLib PUBLIC ./src/)
 
+target_precompile_headers(MainLoopLib PUBLIC ${PROJECT_SOURCE_DIR}/src/common/pch.h)
+
```
Comment (jonapap) [inline CMakeLists.txt:79]: Cool! Didn't know this was a thing. Also, where did you learn all those tricks?

File: src/IO/IO.cpp [hunk 3]
```diff
@@ -10,7 +10,7 @@ IO::IO()
 IO::~IO()
 {
     // destroy thread
-    SPDLOG_LOGGER_DEBUG("destroy thread");
+    SPDLOG_LOGGER_DEBUG(logger, "destroy thread");
     thisThread.~thread(); // thread not killed
     SPDLOG_LOGGER_DEBUG(logger, "Thread testSensorThread killed:");
```
Comment (jonapap) [inline src/IO/IO.cpp:15]:
```suggestion
    SPDLOG_LOGGER_DEBUG(logger, "Destroying thread");
    thisThread.~thread();
```

While we are here, let's fix those horrible messages and comments

File: CMakeLists.txt (hunk 2 above)
Comment (SeoFernando25) [reply CMakeLists.txt:79]: The first time I've heard about it was in [this video](https://youtu.be/eSI4wctZUto?t=1090), his C++ videos are pretty neat 👌

File: src/common/pch.h (hunk 1 above)
Comment (jonapap) [reply src/common/pch.h:19]: Just realized that making `pch.h` not a precompiled header gives a weird warning about `SPDLOG_ACTIVE_LEVEL` being redefined. However, now that logging seems to be properly working, this warning probably isn't worth our time to look at.

**Pull request merged**